frontend/app/src/store/history/index.ts

```diff
@@ -1,17 +1,916 @@
+      const location = parameters?.location ?? '';
+      const exchange = location
+        ? exchangeName(location as TradeLocation)
+        : i18n.tc('actions.trades.all_exchanges');
+      const taskMeta = {
+        title: i18n.tc('actions.trades.task.title'),
+        description: i18n.tc('actions.trades.task.description', undefined, {
+          exchange
+        }),
+        location,
+        numericKeys: []
+      };
+
+      const { result } = await awaitTask<
+        CollectionResponse<EntryWithMeta<Trade>>,
+        TaskMeta
+      >(taskId, taskType, taskMeta, true);
+
+      setStatus(
+        isTaskRunning(taskType).value ? Status.REFRESHING : Status.LOADED
+      );
+
+      const parsedResult = TradeCollectionResponse.parse(result);
+      return mapCollectionEntriesWithMeta<Trade>(
+        mapCollectionResponse(parsedResult)
+      ) as Collection<TradeEntry>;
+    };
+
+    try {
+      const firstLoad = isFirstLoad();
+      const onlyCache = firstLoad ? false : !refresh;
+      if ((isTaskRunning(taskType).value || loading()) && !onlyCache) {
+        return;
+      }
+
+      if (firstLoad || refresh) {
+        fetchAssociatedLocations().then();
+      }
+
+      const fetchOnlyCache = async () => {
+        trades.value = await fetchTradesHandler(true);
+      };
+
+      setStatus(firstLoad ? Status.LOADING : Status.REFRESHING);
+
+      if (!onlyLocation) await fetchOnlyCache();
+
+      if (!onlyCache || onlyLocation) {
+        setStatus(Status.REFRESHING);
+        const { notify } = useNotifications();
+
+        const exchange = onlyLocation
+          ? exchangeName(onlyLocation as TradeLocation)
+          : i18n.tc('actions.trades.all_exchanges');
+
+        await fetchTradesHandler(false, { location: onlyLocation }).catch(
+          error => {
+            notify({
+              title: i18n.tc('actions.trades.error.title', undefined, {
+                exchange
+              }),
+              message: i18n.tc('actions.trades.error.description', undefined, {
+                exchange,
+                error
+              }),
+              display: true
+            });
+          }
+        );
+
+        if (!onlyLocation) await fetchOnlyCache();
+      }
+
+      setStatus(
+        isTaskRunning(taskType).value ? Status.REFRESHING : Status.LOADED
+      );
+    } catch (e) {
+      logger.error(e);
+      setStatus(Status.NONE);
+    }
+  };
+
+  const updateTradesPayload = (newPayload: Partial<TradeRequestPayload>) => {
+    if (!isEqual(tradesPayload.value, newPayload)) {
+      tradesPayload.value = newPayload;
+      fetchTrades().then();
+    }
+  };
+
+  const addExternalTrade = async (trade: NewTrade): Promise<ActionStatus> => {
+    let success = false;
+    let message = '';
+    try {
+      await api.history.addExternalTrade(trade);
+      success = true;
+    } catch (e: any) {
+      message = e.message;
+    }
+
+    fetchAssociatedLocations().then();
+    fetchTrades().then();
```

Comment:
Ideally here `fetchTrades` should run after `fetchAssociatedLocations` without blocking return

Reply:
If the fetchTrades doesn't need the associated locations then we can keep it as it is